### PR TITLE
feat: Sieve of eratosthenes

### PR DESCRIPTION
**Pull request by thechubbypanda**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1730"></a>

### CODE REVIEW

File: math/eratosthenes.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            std::cout << i << std::endl;
+        }
+    }
+}
+
+/**
+ * @brief Self-tests the sieve function for major inconsistencies
+ * @returns void
+ */
+void test() {
```
Comment (Panquesito7) [inline math/eratosthenes.cpp:57]:
```suggestion
static void test() {
```

File: math/eratosthenes.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+ * @author [Keval Kapdee](https://github.com/thechubbypanda)
+ */
+
+#include <cassert>   /// For assert
+#include <chrono>    /// For timing the sieve
+#include <iostream>  /// For IO operations
+#include <string>    /// For string handling
+#include <vector>    /// For std::vector
+
+/**
```
Comment (Panquesito7) [inline math/eratosthenes.cpp:22]:
```suggestion
/**
 * @namespace math
 * @brief Mathematical algorithms
 */
namespace math {
/**
```

File: math/eratosthenes.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+ * @param primes The vector that has been passed through `sieve(...)`
+ * @returns void
+ */
+void print_primes(std::vector<bool> const& primes) {
+    for (int i = 0; i < primes.size(); i++) {
+        if (primes[i]) {
+            std::cout << i << std::endl;
+        }
+    }
+}
```
Comment (Panquesito7) [inline math/eratosthenes.cpp:51]:
```suggestion
}
}  // namespace math
```

File: math/eratosthenes.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+/**
+ * @brief Main function
+ * @param argc commandline argument count
+ * @param argv commandline array of arguments
+ * @returns 0 on exit
+ */
+int main(int argc, char* argv[]) {
+    test();
```
Comment (Panquesito7) [inline math/eratosthenes.cpp:79]:
```suggestion
    test();  // run self-test implementations
```

File: math/eratosthenes.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    }
+}
+
+/**
+ * @brief Self-tests the sieve function for major inconsistencies
+ * @returns void
+ */
+void test() {
+    auto primes = std::vector<bool>(10, true);
+    sieve(&primes);
```
Comment (Panquesito7) [inline math/eratosthenes.cpp:59]:
```suggestion
    math::sieve(&primes);
```

File: math/eratosthenes.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    // Store a boolean for every number which states if that index is prime or
+    // not
+    auto primes = std::vector<bool>(max, true);
+
+    // Store the algorithm start time
+    auto start = std::chrono::high_resolution_clock::now();
+
+    // Run the sieve
+    sieve(&primes);
```
Comment (Panquesito7) [inline math/eratosthenes.cpp:92]:
```suggestion
    math::sieve(&primes);
```

File: math/eratosthenes.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+    // Time difference calculation
+    auto time = std::chrono::duration_cast<
+                    std::chrono::duration<double, std::ratio<1>>>(
+                    std::chrono::high_resolution_clock::now() - start)
+                    .count();
+
+    // Print the primes if we see that "print" was passed as an arg
+    if (argc > 1 && argv[1] == std::string("print")) {
+        print_primes(primes);
```
Comment (Panquesito7) [inline math/eratosthenes.cpp:102]:
```suggestion
        math::print_primes(primes);
```

**Review by Panquesito7 — COMMENTED** [review]
Almost there! 😄

File: math/eratosthenes.cpp [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    assert(primes[0] == false);
+    assert(primes[1] == false);
+    assert(primes[2] == true);
+    assert(primes[3] == true);
+    assert(primes[4] == false);
+    assert(primes[5] == true);
+    assert(primes[6] == false);
+    assert(primes[7] == true);
+    assert(primes[8] == false);
+    assert(primes[9] == false);
```
Comment (Panquesito7) [inline math/eratosthenes.cpp:75]:
```suggestion
    assert(primes[9] == false);

    std::cout << "All tests have successfully passed!\n";
```

**Review by Panquesito7 — DISMISSED** [review]
Awesome work! You've done amazing work, and it's been your
first contribution here! We hope you keep contributing! 😄👍🎉

After this PR is merged, you can ask for the
contributor role in our [**Discord server**](https://discord.gg/c7MnfGFGa6)! 🚀

**Review by ayaankhan98 — CHANGES_REQUESTED** [review]
```
test.cpp: In function ‘void math::sieve(std::vector<bool>*)’:
test.cpp:38:23: error: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<bool>::size_type’ {aka ‘long unsigned int’} [-Werror=sign-compare]
   38 |     for (int n = 2; n < vec->size(); n++) {
      |                     ~~^~~~~~~~~~~~~
test.cpp:39:46: error: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<bool>::size_type’ {aka ‘long unsigned int’} [-Werror=sign-compare]
   39 |         for (int multiple = n << 1; multiple < vec->size(); multiple += n) {
      |                                     ~~~~~~~~~^~~~~~~~~~~~~
test.cpp: In function ‘void math::print_primes(const std::vector<bool>&)’:
test.cpp:51:23: error: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<bool>::size_type’ {aka ‘long unsigned int’} [-Werror=sign-compare]
   51 |     for (int i = 0; i < primes.size(); i++) {
      |                     ~~^~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors

shell returned 1
```

**Review by Panquesito7 — APPROVED** [review]
Awesome work! You've done amazing work, and it's been your
first contribution here! We hope you keep contributing! 😄👍🎉

After this PR is merged, you can ask for the
contributor role in our [**Discord server**](https://discord.gg/c7MnfGFGa6)! 🚀

**Comment by Panquesito7** [comment]
Could you please review again, @ayaankhan98? Thanks. 😄

**Comment by thechubbypanda** [comment]
@Panquesito7 What's the deal with this?

**Review by mishraabhinn — APPROVED** [review]

**Comment by realstealthninja** [comment]
should we close this?

**Comment by Panquesito7** [comment]
Yep, already done in #2442.

**Pull request closed without merging**